drivers/can/can_net.c

```diff
@@ -0,0 +1,254 @@
+
+static inline int attach_unicast_filter(struct net_can_context *ctx)
+{
+	struct zcan_filter filter = {
+		.id_type = CAN_EXTENDED_IDENTIFIER,
+		.rtr = CAN_DATAFRAME,
+		.rtr_mask = 1,
+		.ext_id_mask = CAN_NET_IF_ADDR_DEST_MASK
+	};
+	const u8_t *link_addr = net_if_get_link_addr(ctx->iface)->addr;
+	const u16_t dest = sys_be16_to_cpu(UNALIGNED_GET((u16_t *) link_addr));
+	int filter_id;
+
+	filter.ext_id = (dest << CAN_NET_IF_ADDR_DEST_POS);
+
+	filter_id = can_attach_isr(ctx->can_dev, net_can_recv,
+				   ctx, &filter);
+	if (filter_id == CAN_NET_FILTER_NOT_SET) {
+		NET_ERR("Can't attach FF filter.");
+		return CAN_NET_FILTER_NOT_SET;
+	}
+
+	NET_DBG("Attached FF filter %d.", filter_id);
+
+	return filter_id;
+}
+
+static inline int attach_mcast_filter(struct net_can_context *ctx)
+{
+	struct zcan_filter filter = {
+		.id_type = CAN_EXTENDED_IDENTIFIER,
+		.rtr = CAN_DATAFRAME,
+		.rtr_mask = 1,
+		.ext_id_mask = CAN_NET_IF_ADDR_DEST_MASK
+	};
+	int filter_id;
+
+	filter.ext_id = (NET_CAN_MULTICAST_ADDR << CAN_NET_IF_ADDR_DEST_POS);
+
+	filter_id = can_attach_isr(ctx->can_dev, net_can_recv,
+				   ctx, &filter);
+	if (filter_id == CAN_NET_FILTER_NOT_SET) {
+		NET_ERR("Can't attach multicast filter.");
+		return CAN_NET_FILTER_NOT_SET;
+	}
+
+	NET_DBG("Attached multicast filter %d.", filter_id);
+
+	return filter_id;
+}
+
+static int enable(struct device *dev, bool enable)
+{
+	struct net_can_context *ctx = dev->driver_data;
+
+	if (enable) {
+		if (ctx->recv_filter_id == CAN_NET_FILTER_NOT_SET) {
+			ctx->recv_filter_id = attach_unicast_filter(ctx);
+			if (ctx->recv_filter_id < 0) {
+				return -EIO;
+			}
+		}
+
+		if (ctx->mcast_filter_id == CAN_NET_FILTER_NOT_SET) {
+			ctx->mcast_filter_id = attach_mcast_filter(ctx);
+			if (ctx->mcast_filter_id < 0) {
+				can_detach(ctx->can_dev, ctx->recv_filter_id);
+				return -EIO;
+			}
+		}
+	} else {
+		if (ctx->recv_filter_id != CAN_NET_FILTER_NOT_SET) {
+			can_detach(ctx->can_dev, ctx->recv_filter_id);
+		}
+
+		if (ctx->mcast_filter_id != CAN_NET_FILTER_NOT_SET) {
+			can_detach(ctx->can_dev, ctx->mcast_filter_id);
+		}
+	}
+
+	return 0;
+}
+
+static struct net_can_api net_can_api_inst = {
+	.iface_api.init = net_can_iface_init,
+
+	.send = net_can_send,
+	.attach_filter = attach_filter,
+	.detach_filter = detach_filter,
+	.enable = enable,
+};
+
+static int net_can_init(struct device *dev)
+{
+	struct device *can_dev = device_get_binding(DT_CAN_1_NAME);
+	struct net_can_context *ctx = dev->driver_data;
+
+	ctx->recv_filter_id = CAN_NET_FILTER_NOT_SET;
+	ctx->mcast_filter_id = CAN_NET_FILTER_NOT_SET;
+
+	if (!can_dev) {
+		NET_ERR("Can't get binding to CAN device %s", DT_CAN_1_NAME);
+		return -EIO;
+	}
+
+	NET_DBG("Init net CAN device %p (%s) for dev %p (%s)",
+		dev, dev->config->name, can_dev, can_dev->config->name);
+
+	ctx->can_dev = can_dev;
+
+	return 0;
+}
+
+static struct net_can_context net_can_context_1;
+
+NET_DEVICE_INIT(net_can_stm32_1, CONFIG_CAN_NET_NAME, net_can_init,
```

Comment:
The name of the device  (1st parameter) looks wrong as it contains stm32, what about just calling it "net_can_1"?

Reply:
Copy n paste error.